src/amltk/pytorch/builders.py

```diff
@@ -0,0 +1,167 @@
+"""This module contains functionality to construct a pytorch model from a pipeline.
+
+It also includes classes for handling dimension matching between layers.
+"""
+
+from __future__ import annotations
+
+from torch import nn
+
+from amltk import Choice, Fixed, Sequential
+from amltk.exceptions import MatchChosenDimensionsError, MatchDimensionsError
+
+
+class MatchDimensions:
+    """Handles matching dimensions between layers in a pipeline.
+
+    This class helps ensure compatibility between layers with search spaces
+    during HPO optimization. It takes the layer name and parameter name
+    and stores them for later reference.
+
+    Not intended to be used inside a Choice node.
+
+    Attributes:
+        layer_name (str): The name of the layer.
+        param (str | None): The name of the parameter.
+    """
+
+    def __init__(self, layer_name: str, param: str | None):
+        """Initializes the MatchDimensions object.
+
+        Args:
+           layer_name (str): The name of the layer.
+           param (str | None): The name of the parameter.
+        """
+        self.layer_name = layer_name
+        self.param = param
+
+    def evaluate(self, pipeline: Sequential) -> int:
+        """Retrieves the corresponding configuration value from the pipeline.
+
+        Args:
+            pipeline: The pipeline to search for the matching configuration.
+
+        Returns:
+            The value of the matching configuration parameter.
+        """
+        layer = pipeline[self.layer_name]
+        layer_config = layer.config
+        if layer_config is None:
+            raise MatchDimensionsError(self.layer_name)
+        value = layer_config.get(self.param)
+        if value is None:
```

Comment:
This is a good use case of a `Sentinal Pattern`, i.e. when `None` is a valid value. I don't see many cases for this here but to make configs as flexible as possible, I'm trying to leave `None` as a valid value.

```python
_sentinal = object()
value = layer_config.get(self.param, _sentinal)
if value is _sentinal:
   raise ...
   
# Or without it
if self.param not in lay_config:
   raise ... 
```